misc/toml-parser/src/main/java/io/ballerina/toml/validator/SchemaValidator.java

```diff
@@ -0,0 +1,232 @@
+                            DiagnosticSeverity.ERROR, String.format("Key \"%s\" expects %s . Found array", this.key,
+                                    schema.type()));
+            tomlTableArrayNode.addDiagnostic(diagnostic);
+        } else {
+            ArraySchema arraySchema = (ArraySchema) schema;
+            AbstractSchema items = arraySchema.items();
+            List<TomlTableNode> children = tomlTableArrayNode.children();
+            for (TomlTableNode child : children) {
+                this.schema = items;
+                visit(child);
+            }
+        }
+    }
+
+    @Override
+    public void visit(TomlKeyValueNode keyValue) {
+        TomlValueNode value = keyValue.value();
+        lookupNode(value);
+    }
+
+    @Override
+    public void visit(TomlValueNode tomlValue) {
+        lookupNode(tomlValue);
+    }
+
+    @Override
+    public void visit(TomlStringValueNode tomlStringValueNode) {
+        if (schema.type() != Type.STRING) {
+            TomlDiagnostic diagnostic =
+                    getTomlDiagnostic(tomlStringValueNode.location(), "TVE0002", "error.invalid.type",
+                            DiagnosticSeverity.ERROR,
+                            String.format("Key \"%s\" expects %s . Found string", this.key, schema.type()));
+            tomlStringValueNode.addDiagnostic(diagnostic);
+        } else {
+            StringSchema stringSchema = (StringSchema) this.schema;
+            if (stringSchema.pattern().isPresent()) {
+                String pattern = stringSchema.pattern().get();
+                if (!Pattern.compile(pattern).matcher(tomlStringValueNode.getValue()).matches()) {
+                    TomlDiagnostic diagnostic = getTomlDiagnostic(tomlStringValueNode.location(), "TVE0003",
+                            "error.regex.mismatch", DiagnosticSeverity.ERROR,
+                            String.format("Key \"%s\" value does not match the Regex provided in Schema %s", this.key,
+                                    pattern));
+                    tomlStringValueNode.addDiagnostic(diagnostic);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void visit(TomlDoubleValueNodeNode tomlDoubleValueNodeNode) {
+        if (schema.type() != Type.NUMBER) {
+            TomlDiagnostic diagnostic = getTomlDiagnostic(tomlDoubleValueNodeNode.location(), "TVE0002",
+                    "error.invalid.type", DiagnosticSeverity.ERROR,
+                    String.format("Key \"%s\" expects %s . Found number", this.key, schema.type()));
+            tomlDoubleValueNodeNode.addDiagnostic(diagnostic);
+        } else {
+            List<Diagnostic> diagnostics =
+                    validateMinMaxValues((NumericSchema) schema, tomlDoubleValueNodeNode.getValue(),
+                            tomlDoubleValueNodeNode.location());
+            for (Diagnostic diagnostic : diagnostics) {
+                tomlDoubleValueNodeNode.addDiagnostic(diagnostic);
+            }
```

Comment:
Here we are traversing the diagnostics again.
Instead, can't we add the diagnostics to the node itself directly?